source/Camera/Basic/C++Halcon/CaptureViaZivid/CaptureViaZivid.cpp

```diff
@@ -0,0 +1,88 @@
+#include <Zivid/Zivid.h>
+#include <halconcpp/HalconCpp.h>
+
+#include <chrono>
+#include <iostream>
+
+
+void savePointCloud(const HalconCpp::HObjectModel3D &model, const std::string &filename)
+{
+	model.WriteObjectModel3d(HalconCpp::HString{ "ply" },
+		HalconCpp::HString{ filename.c_str() },
+		HalconCpp::HString{ "invert_normals" },
+		HalconCpp::HString{ "false" });
+}
+
+
+HalconCpp::HObjectModel3D zividToHalconPointCloud(const Zivid::PointCloud &pointCloud)
+{
+	const auto width = pointCloud.width();
+	const auto height = pointCloud.height();
+
+	std::vector<float> array1dX(width * height, 0.0);
+	std::vector<float> array1dY(width * height, 0.0);
+	std::vector<float> array1dZ(width * height, 0.0);
+	std::vector<float> array1dR(width * height, 0.0);
+	std::vector<float> array1dG(width * height, 0.0);
+	std::vector<float> array1dB(width * height, 0.0);
+
+	for (size_t i = 0; i < width*height; ++i)
+	{
+		const auto &point = pointCloud(i);
+		if (!isnan(point.x))
+		{
+			array1dX[i] = point.x;
+			array1dY[i] = point.y;
+			array1dZ[i] = point.z;
+			array1dR[i] = point.red();
+			array1dG[i] = point.green();
+			array1dB[i] = point.blue();
+		}
+	}
+
+	std::cout << "Constructing ObjectModel3D based on XYZ data" << std::endl;
+	const auto X = HalconCpp::HImage{ "real", static_cast<long>(width), static_cast<long>(height), array1dX.data() };
+	const auto Y = HalconCpp::HImage{ "real", static_cast<long>(width), static_cast<long>(height), array1dY.data() };
+	const auto Z = HalconCpp::HImage{ "real", static_cast<long>(width), static_cast<long>(height), array1dZ.data() };
+	HalconCpp::HObjectModel3D objectModel3D(X, Y, Z);
+
```

Comment:
Where did this disappear?

`		// Extract points based on a z-value threshold. Values less than 0.1 or larger than 10000 are discarded. 
		const auto reducedObjectModel3D = objectModel3D.SelectPointsObjectModel3d("point_coord_z", 0.1, 15000);`

Reply:
MVTec explained that instead of selecting points larger than 0.1, it is better to create xyz mappings and then use `set_object_model_3d_attrib_mod` to set the xyz mappings.

This is their explanation:

First of all we need to determine the width of the XYZ mappings. Knowing the width of the xyz mappings, we have an unique correspondence between array indices (for example X[50] as the 50'th index of the array X) and the row and column coordinates of the xyz mappings. How can we define this correspondence?
Assuming our xyz images have a resolution of 520 x 520, i.e width = 520. This means that the first row of the image contains 520 pixels, the second row of the image contains 520 pixel and so on. Therefore, the first 520 indices of the array X are stored at the first row of the X-image. This means that the array index X[50] has the row-column coordinates (0,50) (noting that we start to count at zero), the index X[519] has the row-column coordinate (0,519) and therefore X[520] is then on the next row, having the row-column coordinates (1,0). This means we can easily compute the column coordinate by using the modulo operator:
column = index_x % 520
Furthermore, we can compute the row coordinate by dividing the array index by 520 and then taking the floor function of this value which gives us the maximal integer value less than the input value (e.g |2.5| = 2):
row = |_ index_x / width _|
So, this is how we may compute the xyz images, before we do so, we should check whether the input 3D points are valid or not. The following code snippet shows how this may be done:
for (int i=0; i<width*height; ++i)
```
{
  if (pt_in[i].x != 0 || pt_in[i].y != 0 || pt_in[i].z != 0)
  {
    // This point is valid
    x_out[num_pt_out] = pt_in[i].x;
    y_out[num_pt_out] = pt_in[i].y;
    z_out[num_pt_out] = pt_in[i].z;
    row_out[num_pt_out] = i / width;
    col_out[num_pt_out] = i % width;
    num_pt_out++;
  }
}
```

This means that in for loop (line 29) we want to create row and col arrays. Have in mind that this approach requires changing the indexes of the 1d arrays - x, y, z arrays lose organization (but get smaller for the number of invalid points) but row and col compensate from that.

Reply:
So this loop, which seems to lay grounds to redefine the ObjectModel3D, is better (faster?) than `objectModel3D.SelectPointsObjectModel3d("point_coord_z", 0.1, 15000);`? Did they explain why? Is it for memory reasons? That `objectModel3D.SelectPointsObjectModel3d("point_coord_z", 0.1, 15000);` doesn't actually crop, but somehow labels data?

Reply:
The problem with SelectPintsObjectModel3d is that it creates a new object. We don't know what is faster. This is what they say:
_There should be no real difference between the two methods regarding the run time, but I would normally prefer to do such array operations in the native language. We do not have a real benchmark, but we in the past we have observed that especially when it comes to mathematical computation in a loop, it is often faster to to this by using the native arrays.
Furthermore you have the advantage that no new object model needs to be created._

Reply:
> Where did this disappear?
> 
> ` // Extract points based on a z-value threshold. Values less than 0.1 or larger than 10000 are discarded. const auto reducedObjectModel3D = objectModel3D.SelectPointsObjectModel3d("point_coord_z", 0.1, 15000);`

I removed it when I removed the double for-loop. In the double for-loop I sat every NaN point to z=0. That is why I needed to threshold the point cloud based on z values. It seems that I don't need to handle NaN values myself (Halcon does it for me). I am able to capture and save point clouds without handling this, hence no need to keep the selectPointsObjectModel3d function.

Reply:
From the code below, it seems that you don't assign any values to the 1d arrays when the points are NaN (there is no 'else' that does something else). So it is not Halcon that deals with NaNs, NaNs are ignored.

`if (!isnan(point.x))
		{
			array1dX[i] = point.x;`

I assume these array values automatically get assigned to zeros. But are we certain that they do? If they do, you still have many points in the point cloud with (0,0,0) coordinates. Try to open the saved .ply in mesh lab and you will see if there is a point at (0,0,0).

What I am afraid of is that by not assigning anything to those locations you leave a possibility to get whatever is in memory, which would not be good. Maybe this is can not happen with C++ working with CPU memory, (maybe the memory is automatically flushed when allocated certain space?) I don't know. I know when working with microcontrollers in C this can happen so it is important to either initialize with zeros and then assign nonzeros only in a loop, or assign both zeros and nonzeros in a loop. Not assigning is dangerous. Try to clarify this with someone who understands how memory is handled in C++.

In any case, I think we should implement this as MVTec suggests, that is, with col and row xyz mappings.

Reply:
The total point cloud have 2.3 million points, and indeed there are a "single" point in (0,0,0). So I guess that Halcon sets all NaN points to (0,0,0). 
 
Implementing MVTec suggestions seems like a smart thing to do

Reply:
Done. Without any helpful error messages, this took way longer time than expected... Now Nan values are not included in the halcon point cloud.

Reply:
Great!